### PR TITLE
Lh7051 backingimage encryption support

### PR DESCRIPTION
**Pull request by ChanYiLin**
https://github.com/longhorn/longhorn/issues/7051

Add document for backing image encryption

### CODE REVIEW

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+> - `backing-image`: Specify the source backing image for cloning.
+> - `encryption`: Set the value to `decrypt` to decrypt the backing image during cloning.
+> - `secret`: Specify the secret used to decrypt the backing image.
+> - `secret-namespace`: Specify the namespace of the secret used to decrypt the backing image.
+
+
+## Using an encrypted Volume with the encrypted Backing Image
+
+To use the encrypted volume with the encrypted backing image, the secret for encryption should be the same.
+Once the encrypted backing image is ready, you can create the storage class with the corresponding backing image and the secret to create the volume for the workload.
```
Comment (jillian-maroket) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:168]:
```suggestion
## Use an Encrypted Backing Image with an Encrypted Volume

The secret used to encrypt the backing image and the volume must be identical. Once the encrypted backing image is ready, you can create the StorageClass with the corresponding backing image and the secret to create the volume for the workload. 
```

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+stringData:
+  CRYPTO_KEY_VALUE: "Your encryption passphrase"
+  CRYPTO_KEY_PROVIDER: "secret"
+  CRYPTO_KEY_CIPHER: "aes-xts-plain64"
+  CRYPTO_KEY_HASH: "sha256"
+  CRYPTO_KEY_SIZE: "256"
+  CRYPTO_PBKDF: "argon2i"
+```
+
+Example of YAML code for the storage class:
```
Comment (jillian-maroket) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:187]:
```suggestion
Example of YAML code for the StorageClass:
```

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  # global secret that contains the encryption key that will be used for all volumes
+  csi.storage.k8s.io/provisioner-secret-name: "longhorn-crypto"
+  csi.storage.k8s.io/provisioner-secret-namespace: "longhorn-system"
+  csi.storage.k8s.io/node-publish-secret-name: "longhorn-crypto"
+  csi.storage.k8s.io/node-publish-secret-namespace: "longhorn-system"
+  csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
+  csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
+```
+
+For more details of encrypted volume, please see [volume encryption](../../security/volume-encryption)
```
Comment (jillian-maroket) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:211]:
```suggestion
For more information, see [Volume Encryption](../../security/volume-encryption).
```

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+## Limitation
+
+- We don't support encrypt an encrypted backing image.
+- We don't support decrypt an unencrypted backing image.
+- We don't support change the encryption key for an encrypted backing image.
+- when encrypting qcow2 image, we first temporarily create a raw image from the qcow2 image and then encrypted it. The result will be an encrypted raw image. It will temporarily cost extra raw image size during cloning.
+- If the source backing image is a sparse file, it will loss the sparsity afte encryption.
+- To store the LUKS metadata size when encryption, we expand the image size with extra 16MB, please see [cryptsetup release note](https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/docs/v2.1.0-ReleaseNotes#L27)
```
Comment (jillian-maroket) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:220]:
```suggestion
## Limitations

- Longhorn is unable to encrypt backing images that are already encrypted, and decrypt backing images that are not encrypted.
- Longhorn does not allow you to change the encryption key of an encrypted backing image.
- When encrypting a qcow2 image, Longhorn first creates a raw image from the qcow2 image and then encrypts it. The resulting encrypted raw image temporarily consumes extra space during cloning.
- If the source backing image is a sparse file, the file loses its sparsity after encryption.
- To allow storage of the LUKS metadata during encryption, the image size is increased by 16 MB. For more information, see the [cryptsetup release notes](https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/docs/v2.1.0-ReleaseNotes#L27).
```

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image.md [hunk 5]
```diff
@@ -17,6 +17,7 @@ You can prepare a backing image using four different kinds of data sources.
 2. Upload a file from your local machine. This option is available to Longhorn UI users.
 3. Export an existing in-cluster volume as a backing image.
 4. Restore a backing image from the backupstore, For more information, see [Backing Image Backup](../backing-image-backup).
+5. Clone a backing image from another backing image.
```
Comment (jillian-maroket) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image.md:20]:
```suggestion
5. Clone a backing image.
```

**Review by jillian-maroket — COMMENTED** [review]
Let me know if you have concerns about the changes.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
+---
+title: Backing Image Encryption
+weight: 2
+---
+
+As of v1.7.0, Longhorn supports encrypting and decrypting a backing image through cloning the backing image.
+Backing image encryption support utilizes the linux kernel module `dm_crypt` via `cryptsetup`.
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:7]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+## Clone a Backing Image
+You can clone a backing image using the YAML.
+
+Example of a download backing image:
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:12]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+  name: parrot
+  namespace: longhorn-system
+spec:
+  sourceType: download
+  sourceParameters:
+    url: https://longhorn-backing-image.s3-us-west-1.amazonaws.com/parrot.raw
+  checksum: 304f3ed30ca6878e9056ee6f1b02b328239f0d0c2c1272840998212f9734b196371560b3b939037e4f4c2884ce457c2cbc9f0621f4f5d1ca983983c8cdf8cd9a
+```
+
+Example of YAML code used to clone the sample backing image to a new backing image:
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:27]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+spec:
+  sourceType: clone
+  sourceParameters:
+    backing-image: parrot
+    encryption: ignore
+```
+
+> **IMPORTANT:**
+> - `backing-image`: Specify the source backing image for cloning.
+> - `encryption`: Set the value to `ignore` to directly clone the backing image.
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:44]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+## Encrypt a Backing Image
+You can encrypt a backing image through cloning so the backing image can be used with an encrypted volume.
+
+Example of a download backing image:
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:49]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+  name: parrot
+  namespace: longhorn-system
+spec:
+  sourceType: download
+  sourceParameters:
+    url: https://longhorn-backing-image.s3-us-west-1.amazonaws.com/parrot.raw
+  checksum: 304f3ed30ca6878e9056ee6f1b02b328239f0d0c2c1272840998212f9734b196371560b3b939037e4f4c2884ce457c2cbc9f0621f4f5d1ca983983c8cdf8cd9a
+```
+
+Example of YAML code used to clone and encrypt the sample backing image to a new encrypted backing image:
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:64]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+  CRYPTO_KEY_VALUE: "Your encryption passphrase"
+  CRYPTO_KEY_PROVIDER: "secret"
+  CRYPTO_KEY_CIPHER: "aes-xts-plain64"
+  CRYPTO_KEY_HASH: "sha256"
+  CRYPTO_KEY_SIZE: "256"
+  CRYPTO_PBKDF: "argon2i"
+```
+
+> **IMPORTANT:**
+> - `backing-image`: Specify the source backing image for cloning.
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:99]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+  sourceType: clone
+  sourceParameters:
+    backing-image: parrot-cloned-encrypted
+    encryption: decrypt
+    secret: longhorn-crypto
+    secret-namespace: longhorn-system
+```
+
+> **IMPORTANT:**
+> - `backing-image`: Specify the source backing image for cloning.
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:159]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 1 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:168]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 2 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:187]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 3 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:211]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 4 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:220]: fixed.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image.md (hunk 5 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image.md:20]: fixed.

**Review by jillian-maroket — APPROVED** [review]

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+  csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
+  csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
+```
+
+For more information, see [Volume Encryption](../../security/volume-encryption).
+
+## Limitations
+- Longhorn is unable to encrypt backing images that are already encrypted, and decrypt backing images that are not encrypted.
+- Longhorn does not allow you to change the encryption key of an encrypted backing image.
+- When encrypting a qcow2 image, Longhorn first creates a raw image from the qcow2 image and then encrypts it. The resulting encrypted raw image temporarily consumes extra space during cloning.
```
Comment (derekbit) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:213]: Can you provide an example explaining the maximum extra space usage while cloning? Thank you.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 15]
```diff
@@ -0,0 +1,215 @@
+---
+title: Backing Image Encryption
+weight: 2
+---
+
+Starting with v1.7.0, Longhorn allows you to encrypt and decrypt a backing image by cloning it. The backing image encryption mechanism utilizes the Linux kernel module `dm_crypt` and the command-line utility `cryptsetup`.
+
+## Clone a Backing Image
+You can clone a backing image using YAML code.
```
Comment (derekbit) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:9]: It is not a zero-copy clone, right? It is better to mention a cloned image also consumes disk space.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+apiVersion: longhorn.io/v1beta2
+kind: BackingImage
+metadata:
+  name: parrot-cloned
+  namespace: longhorn-system
+spec:
+  sourceType: clone
+  sourceParameters:
+    backing-image: parrot
+    encryption: ignore
```
Comment (derekbit) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:38]: We probably can improve that `encryption` is automatically mutated with `ignore` if it is an empty string. WDYT?

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 14 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:213]: sure

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 16 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:38]: Agree, I will create a hot fix PR.

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 15 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md]: Yes, added

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+  csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
+```
+
+For more information, see [Volume Encryption](../../security/volume-encryption).
+
+## Limitations
+- Longhorn is unable to encrypt backing images that are already encrypted, and decrypt backing images that are not encrypted.
+- Longhorn does not allow you to change the encryption key of an encrypted backing image.
+- When encrypting a qcow2 image, Longhorn first creates a raw image from the qcow2 image and then encrypts it. The resulting encrypted raw image temporarily consumes extra space during cloning. For example, 
+    1. if we encrypt a 10MiB qcow2 image with a virtual size of 200MiB, we first create the raw image from the qcow2 which will consume 200MiB of the space.
```
Comment (derekbit) [inline content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:214]:
```suggestion
    1. If we encrypt a 10MiB qcow2 image with a virtual size of 200MiB, we first create the raw image from the qcow2 which will consume 200MiB of the space.
```

Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md]: fixed, thanks

**Review by derekbit — APPROVED** [review]

File: content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md (hunk 16 above)
Comment (ChanYiLin) [reply content/docs/1.7.0/advanced-resources/backing-image/backing-image-encryption.md:38]: https://github.com/longhorn/longhorn-manager/pull/2956/files

**Pull request merged**